### ISSUE TITLE
Add UDP to unix domain socket

### ISSUE THREAD

**Issue by dzuy65** [diegonehab/luasocket#165]
This is more like a request.  Have you considered addind UDP to unix domain socket?  I'm willing to contribute,

**Comment by diegonehab** [comment]
Sure, we could have unix.tcp() and unix.udp(). How useful is this? I don't mind adding the support. I never get much feedback from the unix domain implementation.

**Comment by dzuy65** [comment]
Since lua socket has tcp, udp, and unix, unix.tcp, it would just round up the features with unix.udp.
I can use it for my application actually.  I have 2 daemons that use unix socket, but they would lock up some times because on could wait for the other which is blocked by tcp socket.  It would be nice to have unix.udp to avoid blocking.

**Comment by diegonehab** [comment]
So the idea is to split unix.c into unixtcp.c and unixudp.c. Then have unix.c init both modules, create a "unix" table, load both tcp() and udp() into the table, and return it. Right?

**Comment by dzuy65** [comment]
I'm not familiar with the code so I can't really comment on it, but I know that unix.c assumes TCP by default.  It would be nice to have a UDP option.  Yes, I think there should be a unix domain socket, then a choice to use TCP or UDP, with TCP as default.

**Comment by diegonehab** [comment]
Right now, unix.c is kind of an experimental beast. It builds into its own library, and I am not sure how people use it. I suppose they run something like 

    local socket = require"socket"
    socket.unix = require"socket.unix"
    u = socket.unix()

This creates a stream socket.

We  should change this to something like

    local socket = require"socket"
    socket.unix = require"socket.unix"
    u1 = socket.unix.tcp()
    u2 = socket.unix.udp()

Each call creating the appropriate type of socket.

The proper way to do this would be to keep in unix.c only the functions that are common to tcp and udp. Then create a unixtcp.c and move the tcp code to it. Create unixudp.c and write udp code into it. Then add a luaopen_unix in unix.c that creates and returns a table with the tcp and udp functions implemented in unixtcp.c and unixudp.c.

Does that make sense?

**Comment by dzuy65** [comment]
Yes, I'm using 

u = socket.unix()

Reading the code, looks like the unix.c code has all the TCP methods, some may be common to UDP.
I'm thining of:

usocket = socket.unit(proto)

where proto can be either "udp" or "tcp", then you'd just add more UDP methods to unix.c without adding new files.

But you know your code best.  I'll go with whatever you design.

**Comment by diegonehab** [comment]
The problem with the proto idea is that the implementation itself will be different, just like the implementation  in udp.c is different from the one in tcp.c. This is mostly because of the presence of buffering in the case of stream and lack of buffering of dgram. Got it?

**Comment by dzuy65** [comment]
Got it, but I'm having trouble understanding how to create an empty socket.unix.  I can implement the methods send, receive, etc., can you help on the module initialization code?

**Comment by diegonehab** [comment]
Sure!

I think the best way to do it is like what you see in luasocket.c does with the table socket. It creates a table and leaves it on top of the stack. Then it does a for loop initializing the submodules. These add stuff to the table as they wish. luasocket.c then simply returns 1 and the user gets the table.

From the user side, it would look like

    local unix = require"socket.unix"

This will work if you put your unix.so inside the socket directory (e.g., /usr/lib/lua/socket/unix.so) you use in your distribution.

**Comment by dzuy65** [comment]
This is my new unix.c implementation following luasocket.c:

static luaL_Reg mod[] = {
    {"tcp", unix_tcp},
    {"udp", unix_udp},
    {NULL, NULL}
};

static luaL_Reg func[] = {
    {NULL,          NULL}
};

UNIX_API int luaopen_socket_unix(lua_State *L) {
    int i;
#if LUA_VERSION_NUM > 501 && !defined(LUA_COMPAT_MODULE)
    lua_newtable(L);
    luaL_setfuncs(L, func, 0);
#else
    luaL_openlib(L, "socket.unix", func, 0);
#endif
    for (i = 0; mod[i].name; i++) mod[i].func(L); 
    return 1;
}

When run, I got:

attempt to index field 'unix' (a function value)

My lua is 5.1 (openwrt does not support 5.2).  How do I initialize unix so that it's a table?

**Comment by diegonehab** [comment]
First, forget all the LUA_VERSION stuff. Clean up your code and make sure it works only on your platform. 5.1 is fine. I can help you make it generic later on.

All you have to do is inside luaopen_socket_unix() is to create a table, place the functions unix_tcp and unix_udp inside the table as fields, and return the table.

What you are doing here is a bit of a mess. You are calling the functions that you should put inside the table...
